### PR TITLE
800 - Show consistent amounts via formatSmart

### LINKED ISSUE #800: Bug: Incoherent amounts
**Description**
It's showing me incoherent amounts, also incoherent precision in how it shows them.

Issues:
- [ ] Different preccision for the different screens/messages. The swap widget shows two decimals for the minimum amount, but 0 in the confirmation
- [ ]  The Waiting for confirmation modal shows another amount. I think it should show the "at least" notation that should match with what you sign and with the confirmation toast
- [ ] The amounts doesn't match. The toast said 4590, but I was expecting 4595.61

**How to Reproduce**
1. Go to CowSwap
2. Input the trade info
3. Send the trade as quick as possible taking screenshot in each step

**Context**
Testing the parawap price, i came across this error.
It fell like the UI lied to me. It tells you one amount, but after you sign is another in the message.

image

**Details**
### Swap widget
Screenshot at Jun 23 16-55-03

### Confirm Swap
Screenshot at Jun 23 16-55-10

### Waiting for confirmation
Screenshot at Jun 23 16-55-20

### Toast success message
Screenshot at Jun 23 16-55-33

### PR DESCRIPTION
**Pull request by W3stside**
Closes #800 

Replaces use of `.toSignificant(X)` with the `dex-js@0.11.0` method `formatSmart` which is extended inside `utils/format` to ignore `thousandsSeparator`

Creates consistent numbers.

### Testing
1. use the app and looks the numbers

### CODE REVIEW

**Review by biocom — APPROVED** [review]
At first sight it looks good to me. What visual changes should we see vs. the other fn/implementation? 2 decimal cut-off? How to visually measure the higher precision?

File: package.json [hunk 1]
```diff
@@ -19,6 +18,7 @@
     "@reach/dialog": "^0.10.3",
     "@reach/portal": "^0.10.3",
     "@react-hook/window-scroll": "^1.3.0",
+    "@reduxjs/toolkit": "^1.5.0",
```
Comment (anxolin) [inline package.json:21]: why did we need this?

File: src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx [hunk 2]
```diff
@@ -109,7 +115,7 @@ SwapModalHeaderProps) {
                 fontWeight={500}
                 color={showAcceptChanges && trade.tradeType === TradeType.EXACT_OUTPUT ? theme.primary1 : ''}
               >
-                {trade.inputAmount.toSignificant(6)}
+                {formatSmart(trade.inputAmount, DEFAULT_PRECISION)}
```
Comment (anxolin) [inline src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx:118]: nit: Maybe we should have our own formatSmart that has default precission by default and delegate? 
This might remove some repetition and also would make it easier to change the logic in all the places at once?

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx [hunk 3]
```diff
@@ -50,6 +51,11 @@ export default function TradeSummary({
   const isExactIn = trade.tradeType === TradeType.EXACT_INPUT
   const slippageAdjustedAmounts = computeSlippageAdjustedAmounts(trade, allowedSlippage)
 
+  const [slippageOut, slippageIn] = useMemo(
+    () => [slippageAdjustedAmounts[Field.OUTPUT], slippageAdjustedAmounts[Field.INPUT]],
+    [slippageAdjustedAmounts]
+  )
+
```
Comment (anxolin) [inline src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:58]: These changes are not related afait, probably u used the wrong base branch?

**Review by anxolin — COMMENTED** [review]
APPROVE (soft)

I did some shallow testing. I would make sure in the tests we ensure that big amounts don't break the UI.

Overall, i'm very happy with this change!!!

File: package.json (hunk 1 above)
Comment (W3stside) [reply package.json:21]: this was automatic from installation, the change is just a line positioning, it didn't update anything

File: package.json [hunk 4]
```diff
@@ -166,6 +166,7 @@
   "license": "GPL-3.0-or-later",
   "dependencies": {
     "@gnosis.pm/cow-runner-game": "^0.2.2",
+    "@gnosis.pm/dex-js": "^0.10.0",
```
Comment (W3stside) [inline package.json]: this is the package update @anxolin

File: src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx (hunk 2 above)
Comment (W3stside) [reply src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx:118]: but we do, take a look inside `utils/format` I am wrapping dex-js formatSmart. I chose to keep the option for precision as it may be useful later

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx (hunk 3 above)
Comment (W3stside) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:58]: no, it was syntactic sugar for making the change a bit easier, before it was using the entire `slippageAdjustedAmounts[Field.OUTPUT]` name and it got annoying. It's not a major change but I can change it back if we don't like it @anxolin @alfetopito

File: src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx (hunk 2 above)
Comment (W3stside) [reply src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx:118]: https://github.com/gnosis/cowswap/pull/895/files#diff-5e3ddbfbe5376cda8fb7e0ae4ac8de168317a2182dbdf93d0445ddec6e302953R18

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx (hunk 3 above)
Comment (alfetopito) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:58]: I'm good, I like the change

**Review by alfetopito — COMMENTED** [review]
Great job, much more smooth that I expected.

Did a test with tokens with a huge price. Not hard to find on Rinkeby.

1. First, the balance. Looks ok and easier to follow IMO.
screenshot_2021-07-08_09-35-08

2. Buying huge amounts:
All good AFACIT
screenshot_2021-07-08_09-45-52
screenshot_2021-07-08_09-45-42
screenshot_2021-07-08_09-36-13

3. Buying tiny amounts
Not so good. I think in the inputs we shouldn't ever show `< x`.
screenshot_2021-07-08_10-02-02
screenshot_2021-07-08_10-02-24
screenshot_2021-07-08_10-02-34
Also in the toast notification, the `<` sign is not present. Not sure why.
screenshot_2021-07-08_10-02-54

For small amounts in this case, we could do like the explorer:
screenshot_2021-07-08_10-03-18
In these cases, I suggest to use the token decimals as the amounts of decimals to display.

File: src/custom/utils/format.ts [hunk 5]
```diff
@@ -1,7 +1,31 @@
...
+  const amount = value instanceof CurrencyAmount ? value.quotient.toString() : value.toFixed(decimalsToShow)
+
+  return _formatSmart({
+    amount,
+    precision,
+    decimals: decimalsToShow,
+    thousandSeparator: false,
+  })
+}
+;(window as any).formatSmart = formatSmart
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/utils/format.ts:31]: oops nope just testing

[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/utils/format.ts:31]:
```suggestion
```

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx (hunk 3 above)
Comment (W3stside) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:58]: 👍

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx [hunk 6]
```diff
@@ -65,7 +71,7 @@ export default function TradeSummary({
           )}
         </RowFixed>
         <TYPE.black textAlign="right" fontSize={12} color={theme.text1}>
-          {realizedFee ? `${realizedFee.toSignificant(DEFAULT_PRECISION)} ${realizedFee.currency.symbol}` : '-'}
+          {`${formatSmart(realizedFee ?? undefined, DEFAULT_PRECISION) || '-'} ${realizedFee?.currency.symbol}`}
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:74]: i wholeheartedly refuse

[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:74]: ok i'll do it

File: src/custom/pages/Swap/SwapMod.tsx [hunk 7]
```diff
@@ -663,7 +665,7 @@ export default function Swap({
                         <Trans>Slippage Tolerance</Trans>
                       </ClickableText>
                       <ClickableText fontWeight={500} fontSize={14} color={theme.text2} onClick={toggleSettings}>
-                        {allowedSlippage.toSignificant(2)}%
+                        {formatSmart(allowedSlippage, 2)}%
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/pages/Swap/SwapMod.tsx:668]: i have FIAT_PRECISION = 2, is that too unclear?

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:74]: :shrug:

File: src/custom/pages/Swap/SwapMod.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (alfetopito) [reply src/custom/pages/Swap/SwapMod.tsx:668]: Yeah, they are different things. I would prefer another const for percentage percision.

File: src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx (hunk 2 above)
Comment (anxolin) [reply src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx:118]: Ahh, right. Still a nit, but if u set `DEFAULT_PRECISION` as the default for `decimalsToShow`, you remove most of the second arguments. 

image

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx (hunk 3 above)
Comment (anxolin) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:58]: okok 👌, I thought it was comming from other PR. NVM

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (anxolin) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx:74]: 😂

**Comment by anxolin** [comment]
> Not so good. I think in the inputs we shouldn't ever show < x.

Yep, I think this one is important to fix.
Comparing with current version, small amounts in PRO looks better.
image

I remember there was a scape hatch for smartFormat. This plus the right precision should do the trick right?

**Comment by alfetopito** [comment]
> > Not so good. I think in the inputs we shouldn't ever show < x.
> 
> Yep, I think this one is important to fix.
> Comparing with current version, small amounts in PRO looks better.
> image
> 
> I remember there was a scape hatch for smartFormat. This plus the right precision should do the trick right?

That would be the `smallLimit` https://docs.google.com/document/d/1PsdWn9WPFf2BNeoJNtClyra8jNvWk_bJ-g33TbpUf8s/edit?disco=AAAANSSWEX8&ts=60e74667&usp=slack

**Comment by W3stside** [comment]
> > Not so good. I think in the inputs we shouldn't ever show < x.
> 
> Yep, I think this one is important to fix.
> Comparing with current version, small amounts in PRO looks better.
> image
> 
> I remember there was a scape hatch for smartFormat. This plus the right precision should do the trick right?

@anxolin @alfetopito 

Fixed: bec15a8

File: src/custom/utils/format.ts [hunk 8]
```diff
@@ -1,7 +1,31 @@
...
+/**
+ * formatSmart
+ * @param amount
+ * @param decimalsToShow
+ * @returns string or undefined
+ */
+export function formatSmart(value: CurrencyAmount<Currency> | Percent | Fraction | undefined, decimalsToShow: number) {
+  if (!value) return
+
+  const precision = value instanceof CurrencyAmount ? value.currency.decimals : 0
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/utils/format.ts:26]: :D

File: src/custom/pages/Swap/SwapMod.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/pages/Swap/SwapMod.tsx]: done

File: src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/swap/TradeSummary/TradeSummaryMod.tsx]: done

File: src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx (hunk 2 above)
Comment (W3stside) [reply src/custom/components/swap/SwapModalHeader/SwapModalHeaderMod.tsx]: @anxolin done 544bac2

File: src/custom/utils/tooltips.ts [hunk 9]
```diff
@@ -1,11 +1,13 @@
...
 export function getMinimumReceivedTooltip(allowedSlippage: Percent, isExactIn: boolean): string {
   const slippagePercent = allowedSlippage
 
   return `${
     isExactIn ? "Minimum tokens you'll receive." : "Maximum tokens you'll sell."
-  } This accounts for the current price and your chosen slippage (${slippagePercent.toSignificant(
-    3
+  } This accounts for the current price and your chosen slippage (${formatSmart(
+    slippagePercent,
+    2
```
Comment (alfetopito) [inline src/custom/utils/tooltips.ts:11]: There's now a const for that ;)

**Review by alfetopito — COMMENTED** [review]
Approved! (just one tiny change left)

File: src/custom/utils/tooltips.ts (hunk 9 above)
Comment (W3stside) [reply src/custom/utils/tooltips.ts:11]: Oops yeah true
